src/schedule/mod.rs

```diff
@@ -1,174 +1,340 @@
-            init_secret,
+        group_context: &GroupContext,
+        member_secret: MemberSecret,
+    ) -> Self {
+        EpochSecret {
+            secret: member_secret.secret.hkdf_expand_label(
+                ciphersuite,
+                "epoch",
+                &group_context.serialize(),
+                ciphersuite.hash_length(),
+            ),
+        }
+    }
+
+    //
+}
+
+/// The `EncryptionSecret` is used to create a `SecretTree`.
+pub struct EncryptionSecret {
+    secret: Secret,
+}
+
+impl EncryptionSecret {
+    /// Derive an encryption secret from a reference to an `EpochSecret`.
+    fn from_epoch_secret(ciphersuite: &Ciphersuite, epoch_secret: &EpochSecret) -> Self {
+        EncryptionSecret {
+            secret: epoch_secret.secret.derive_secret(ciphersuite, "encryption"),
         }
     }
+
     /// Create a `SecretTree` from the `encryption_secret` contained in the
     /// `EpochSecrets`. The `encryption_secret` is replaced with `None` in the
     /// process, allowing us to achieve FS.
-    pub fn create_secret_tree(
-        &mut self,
-        treesize: LeafIndex,
-    ) -> Result<SecretTree, KeyScheduleError> {
-        let encryption_secret = self.consume_encryption_secret()?;
-        Ok(SecretTree::new(encryption_secret, treesize))
-    }
-
-    /// Consume the `encryption_secret` from the `EpochSecrets`, replacing it
-    /// with `None` and return it.
-    fn consume_encryption_secret(&mut self) -> Result<Secret, KeyScheduleError> {
-        let encryption_secret = match self.encryption_secret.take() {
-            Some(es) => es,
-            None => return Err(KeyScheduleError::SecretReuseError),
-        };
-        Ok(encryption_secret)
+    pub fn create_secret_tree(self, treesize: LeafIndex) -> SecretTree {
+        SecretTree::new(self, treesize)
+    }
+
+    pub(crate) fn consume_secret(self) -> Secret {
+        self.secret
+    }
+
+    /// Create a random `EncryptionSecret`. For testing purposes only.
+    #[cfg(test)]
+    pub fn from_random(length: usize) -> Self {
+        EncryptionSecret {
+            secret: Secret::from_random(length),
+        }
+    }
+}
+
+/// A secret that we can derive secrets from, that are used outside of OpenMLS.
+#[derive(Debug)]
+pub(crate) struct ExporterSecret {
+    secret: Secret,
+}
+
+impl ExporterSecret {
+    /// Derive an `ExporterSecret` from an `EpochSecret`.
+    pub(crate) fn from_epoch_secret(ciphersuite: &Ciphersuite, epoch_secret: &EpochSecret) -> Self {
+        let secret = epoch_secret.secret.derive_secret(ciphersuite, "exporter");
+        ExporterSecret { secret }
+    }
+
+    /// Get the `Secret` of the `ExporterSecret`.
+    pub(crate) fn secret(&self) -> &Secret {
+        &self.secret
     }
+}
+
+/// The `EpochSecrets` contain keys (or secrets), which are accessible outside
+/// of the `KeySchedule` and which don't get consumed immediately upon first
+/// use.
+#[derive(Debug)]
+pub struct EpochSecrets {
```

Comment:
I don't see this being used outside of the library.

Reply:
This might be an API problem. Right now, the public function `to_plaintext` takes `EpochSecrets` as input. It's not obvious to me what the best way is to solve this. Should I file an issue?

Reply:
I don't think that `to_plaintext` should be public. It's only used `decrypt`, which should be the public function.